sqlparser/lib/src/analysis/steps/nested_query_resolver.dart

```diff
@@ -0,0 +1,68 @@
+part of '../analysis.dart';
+
+/// Converts all references in nested queries, that require data from the
+/// parent query into variables.
+class NestedQueryResolver extends RecursiveVisitor<void, void> {
+  final AnalysisContext context;
+
+  NestedQueryResolver(this.context);
+
+  @override
+  void visitMoorSpecificNode(MoorSpecificNode e, void arg) {
+    if (e is NestedQueryColumn) {
+      _transform(context, e);
+    } else {
+      super.visitMoorSpecificNode(e, arg);
+    }
+  }
+}
+
+void _transform(AnalysisContext context, NestedQueryColumn e) {
+  e.select.transformChildren(_NestedQueryTransformer(context), null);
+
+  AstPreparingVisitor.resolveIndexOfVariables(
+    e.allDescendants.whereType<Variable>().toList(),
+  );
+}
+
+class _NestedQueryTransformer extends Transformer<void> {
+  final AnalysisContext context;
+
+  _NestedQueryTransformer(this.context);
+
+  @override
+  AstNode? visitReference(Reference e, void arg) {
+    // if the scope of the nested query cannot resolve the reference, the
+    // reference needs to be retrieved from the parent query
+    if (e.entityName != null &&
+        e.scope.resolve(e.entityName!, includeParents: false) == null) {
+      final result = e.scope.resolve(e.entityName!);
+
+      if (result == null) {
+        context.reportError(AnalysisError(
+          type: AnalysisErrorType.referencedUnknownTable,
+          message: 'Unknown table or view in nested query: ${e.entityName}',
+          relevantNode: e,
+        ));
+      } else {
+        return NestedQueryVariable(
+          entityName: e.entityName,
+          columnName: e.columnName,
+        )..setSpan(e.first!, e.last!);
```

Comment:
I'm not sure if I like entirely synthetic AST nodes introduced this late during analysis.

Is it possible to move parts of this analysis into the `drift_dev` package so that `sqlparser` isn't concerned with all of it? We could somehow mark references that need capturing so that the drift-specific rewriter can pick those up later?

Reply:
The problem I ran into and why I left them inside of the sqlparser is, that after introducing these syntactic variables the indices of all variables have to be calculated again (after the transform call in the _transfrom method) with: 
```dart
AstPreparingVisitor.resolveIndexOfVariables(
  e.allDescendants.whereType<Variable>().toList(),
);
```
And if we move this step to drift_dev some kind of `resolveIndexOfVariables` method needs to be exposed.